File: Dynamic Programming/139. Word Break/WordBreak.java
import java.util.HashMap;
import java.util.List;


class Solution {
    HashMap<String, Boolean> map;

    private boolean helper(String s, List<String> wordDict) {
        if(s.isEmpty()) return true;
        if(map.containsKey(s)) return map.get(s);
        for (String word : wordDict) {
            if(s.startsWith(word)) {
                if(helper(s.substring(word.length()), wordDict)) return true;
            }
        }
        map.put(s, false);
        return false;
    }

    public boolean wordBreak(String s, List<String> wordDict) {
        map = new HashMap<>();
        return helper(s, wordDict);
    }
}

File: Dynamic Programming/139. Word Break/WordBreak.py
class Solution(object):

    def helper(self, s, wordDict, d):
        if len(s)==0: return True
        if s in d: return d[s]

        for word in wordDict:
            if s.startswith(word) and self.helper(s[len(word):], wordDict, d):
                return True

        d[s] = False
        return False
        
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        return self.helper(s, wordDict, {})



s = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"
wordDict = ["a","aa","aaa","aaaa","aaaaa","aaaaaa","aaaaaaa","aaaaaaaa","aaaaaaaaa","aaaaaaaaaa"]
print(Solution().wordBreak(s, wordDict))